connectors/src/connectors/slack/temporal/activities.ts

```diff
@@ -0,0 +1,339 @@
+  }
+  const text = await formatNonThreaded(channelId, messages);
+
+  const startDate = new Date(startTsMs);
+  const endDate = new Date(endTsMs);
+  const startDateStr = `${startDate.getFullYear()}-${startDate.getMonth()}-${startDate.getDate()}`;
+  const endDateStr = `${endDate.getFullYear()}-${endDate.getMonth()}-${endDate.getDate()}`;
+  const documentId = `${channelName}-nonthreaded-${startDateStr}-${endDateStr}`;
+  await upsertToDatasource(dataSourceConfig, documentId, text);
+
+  return c;
+}
+
+export async function syncThreads(
+  dataSourceConfig: DataSourceConfig,
+  slackAccessToken: string,
+  channelId: string,
+  threadsTs: string[]
+) {
+  while (threadsTs.length > 0) {
+    const _threadsTs = threadsTs.splice(0, MAX_CONCURRENCY_LEVEL);
+
+    await Promise.all(
+      _threadsTs.map((t) =>
+        syncThread(dataSourceConfig, slackAccessToken, channelId, t)
+      )
+    );
+  }
+}
+
+export async function syncThread(
+  dataSourceConfig: DataSourceConfig,
+  slackAccessToken: string,
+  channelId: string,
+  threadTs: string
+) {
+  const client = new WebClient(slackAccessToken);
+
+  let allMessages: Message[] = [];
+  const { channel } = await client.conversations.info({
+    channel: channelId,
+  });
+  if (!channel) {
+    throw new Error("Channel not found for id " + channelId);
+  }
+
+  let next_cursor = undefined;
+
+  do {
+    const replies: ConversationsRepliesResponse =
+      await client.conversations.replies({
+        channel: channelId,
+        ts: threadTs,
+        cursor: next_cursor,
+      });
+    if (replies.error) {
+      throw new Error(replies.error);
+    }
+    if (!replies.messages) {
+      break;
+    }
+    allMessages = allMessages.concat(replies.messages);
+    next_cursor = replies.response_metadata?.next_cursor;
+  } while (next_cursor);
+
+  const text = await formatThread(channelId, allMessages);
+  const documentId = `${channel.name}-threaded-${threadTs}`;
+
+  await upsertToDatasource(dataSourceConfig, documentId, text);
+}
+
+async function formatThread(channelId: string, messages: Message[]) {
+  const promises = messages.map(async (message) => {
+    const userName = await getUserName(message.user as string);
+    let text = await processMessageForMentions(message.text as string);
+    text = processMessageRemoveEmoji(text);
+
+    return `${userName} said:\n ${text}`;
+  });
+
+  return (await Promise.all(promises)).join("\n");
+}
+
+async function processMessageForMentions(message: string): Promise<string> {
+  const matches = message.match(/<@[A-Z-0-9]+>/g);
+  if (!matches) {
+    return message;
+  }
+  for (const m of matches) {
+    const userId = m.replace(/<|@|>/g, "");
+    const userName = await getUserName(userId);
+    if (!userName) {
+      continue;
+    }
+
+    message = message.replace(m, userName);
+
+    continue;
+  }
+
+  return message;
+}
+
+function processMessageRemoveEmoji(message: string): string {
+  return message.replace(/:[a-z_-]+:/g, "");
```

Comment:
I think we should keep emojis ?

Reply:
I think so. As utf8 or ascii format the semantic content is generally very useful and models are good at parsing it 👍